gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +437,113 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
+   * @return {@link ChannelCredentials} to use to create a plaintext connection between client and
+   *     S2A
+   */
+  ChannelCredentials createPlaintextToS2AChannelCredentials(String plaintextAddress) {
+    if (Strings.isNullOrEmpty(plaintextAddress)) {
+      return null;
+    }
+    return S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
+        .build();
+  }
+
+  /**
+   * This method creates gRPC {@link ChannelCredentials} configured to use S2A to estbalish a mTLS
+   * connection. First, the address of S2A is discovered by using the {@link S2A} utility to learn
+   * the {@code mtlsAddress} to reach S2A and the {@code plaintextAddress} to reach S2A. Prefer to
+   * use the {@code mtlsAddress} address to reach S2A if it is non-empty and the MTLS-MDS
+   * credentials can successfully be discovered and used to create {@link TlsChannelCredentials}. If
+   * there is any failure using mTLS-to-S2A, fallback to using a plaintext connection to S2A using
+   * the {@code plaintextAddress}. If {@code plaintextAddress} is not available, this function
+   * returns null; in this case S2A will not be used, and a TLS connection to the service will be
+   * established.
+   *
+   * @return {@link ChannelCredentials} configured to use S2A to create mTLS connection to
+   *     mtlsEndpoint.
+   */
+  ChannelCredentials createS2ASecuredChannelCredentials() {
+    S2A s2aUtils = S2A.newBuilder().build();
+    String plaintextAddress = s2aUtils.getPlaintextS2AAddress();
+    String mtlsAddress = s2aUtils.getMtlsS2AAddress();
+    if (Strings.isNullOrEmpty(mtlsAddress)) {
+      // Fallback to plaintext connection to S2A.
+      return createPlaintextToS2AChannelCredentials(plaintextAddress);
+    }
+    // Currently, MTLS to MDS is only available on GCE. See:
+    // https://cloud.google.com/compute/docs/metadata/overview#https-mds
+    // Try to load MTLS-MDS creds.
+    File rootFile = new File(MTLS_MDS_ROOT);
+    File certKeyFile = new File(MTLS_MDS_CERT_CHAIN_AND_KEY);
+    if (!rootFile.isFile() || !certKeyFile.isFile()) {
+      // Try to connect to S2A using mTLS.
+      ChannelCredentials mtlsToS2AChannelCredentials = null;
+      InputStream trustBundle = null;
+      InputStream privateKey = null;
+      InputStream certChain = null;
+      try {
+        trustBundle = new FileInputStream(MTLS_MDS_ROOT);
+        privateKey = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+        certChain = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+      } catch (FileNotFoundException ignore) {
+        // Fallback to plaintext-to-S2A connection on error.
+        LOG.log(
+            Level.INFO,
+            "Cannot establish an mTLS connection to S2A due to error loading MTLS to MDS credentials, falling back to plaintext connection to S2A: "
+                + ignore.getMessage());
+        return createPlaintextToS2AChannelCredentials(plaintextAddress);
```

Comment:
If we already validated that all three files exist, then this is a true error scenario that we should throw a runtime exception and log error, instead of still fall back to `PlaintextToS2AChannelCredentials`. But I see that `certChain` is not validated at this moment, is a reason for it?

Reply:
> But I see that certChain is not validated at this moment, is a reason for it?

it is validated as part of `certKeyFile.isFile()`; both the cert and key are stored in MTLS_MDS_CERT_CHAIN_AND_KEY. I included a comment above declaration of MTLS_MDS_CERT_CHAIN_AND_KEY.

>If we already validated that all three files exist, then this is a true error scenario that we should throw a runtime exception and log error, instead of still fall back to PlaintextToS2AChannelCredentials

I agree that this would be an error scenario, so I have changed it to log a warning in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/7d7b2337938df982e9a22621397995eb1a888fbe. However, we do not want to throw a runtime exception because this would cause failing to create a channel if there is a problem with MTLS-MDS. We want to fallback to having a plaintext connection to S2A, because an error here doesn't indicate anything wrong with our flow, it is an error with the MTLS-MDS feature owned by GCE. Eventually, we may (at present, there is not a timeline for this) get rid of the plaintext-to-S2A fallback, in which case this code will be modified to return null on any error when trying to do mtls-to-S2A and fallback to TLS (`builder = ManagedChannelBuilder.forAddress(serviceAddress, port);`) will happen outside this function.

Essentially, we do not want to cause a runtime exception if we cannot use S2A. 

To summarize, the fallback is as follows:
(we have decided to try to use S2A)
- try create S2A channel credentials supporting mTLS-to-S2A:
```
channelCreds = S2AChannelCredentials.newBuilder(mtlsAddress, mtlsToS2AChannelCredentials).build()
builder = Grpc.newChannelBuilder(endpoint, channelCreds); // endpoint is mtls endpoint
```
- fallback to S2A channel credentials supporting plaintext-to-S2A:
```
channelCreds = S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
        .build()
builder = Grpc.newChannelBuilder(endpoint, channelCreds); // endpoint is mtls endpoint
```
- fallback to TLS:
```
builder = ManagedChannelBuilder.forAddress(serviceAddress, port); // serviceAddress is non-mtls endpoint
```

If we eventually stop supporting a plaintext S2A, the fallback will be as follows:
(we have decided to try to use S2A)
- try create S2A channel credentials supporting mTLS-to-S2A:
```
channelCreds = S2AChannelCredentials.newBuilder(mtlsAddress, mtlsToS2AChannelCredentials).build()
builder = Grpc.newChannelBuilder(endpoint, channelCreds); // endpoint is mtls endpoint
```
- fallback to TLS:
```
builder = ManagedChannelBuilder.forAddress(serviceAddress, port); // serviceAddress is non-mtls endpoint
```

WDYT?

Reply:
> both the cert and key are stored in MTLS_MDS_CERT_CHAIN_AND_KEY

Ah, thanks! I didn't notice that both file streams are coming from the same paths. Is it a common practice though, that storing both cert and key in the same file?

> because an error here doesn't indicate anything wrong with our flow, it is an error with the MTLS-MDS feature owned by GCE

I think that makes sense if the files do not exist due to errors in GCE. Is it still possible to run into errors due to "error with the MTLS-MDS feature owned by GCE" if files already exist? Or is it only because we are forced to handle `FileNotFoundException` and we want to fall back to `PlaintextToS2AChannelCredentials` if `FileNotFoundException` happens? If it is the latter, I think we can avoid handling `FileNotFoundException` by passing `File` to `TlsChannelCredentials` directly.

Reply:
>Is it a common practice though, that storing both cert and key in the same file?

I think usually we find cert chain and private key in separate files, since the cert is the public part that you distribute to others. However, the APIs are able to handle if both the cert and key are appearing in the same file.

> If it is the latter,
It is the latter: because we are forced to handle FileNotFoundException and we want to fall back to PlaintextToS2AChannelCredentials if FileNotFoundException happens

Yes, we can avoid handling this case by passing a File to TlsChannelCredentials directly, instead of the InputStream. The reason I had `createMtlsToS2AChannelCredentials` take an InputStream initially was so that I could test by getting the input stream from classpath(could not find a way to get File object from classpath), and I thought getting resources from classpath is preferred:
```
    InputStream trustBundle = this.getClass().getClassLoader().getResourceAsStream("root_cert.pem");
    InputStream privateKey = this.getClass().getClassLoader().getResourceAsStream("client_key.pem");
    InputStream certChain = this.getClass().getClassLoader().getResourceAsStream("client_cert.pem");
```

However, I can do: `File root = new File("src/test/resources/root_cert.pem");` (not loading from classpath) and it works just fine from tests. Let me know if it looks ok in https://github.com/googleapis/sdk-platform-java/pull/3326/commits/a42dcbdf3c2eed8518d14c09dabf737ca99f865b

Reply:
Thanks, LGTM!